### PR TITLE
feat: scex expression summary cube builder

### PR DESCRIPTION
**Pull request by MDunitz**
### Reviewers
**Functional:** 

**Readability:** 

---

## Changes
- add code to build expression summary cube
- update pipeline to call cell count cube builder function

### CODE REVIEW

File: backend/wmg/requirements.txt [hunk 1]
```diff
@@ -79,6 +79,7 @@ requests==2.27.1
 rsa==4.8
 s3transfer==0.5.2
 sarif-om==1.0.4
+scanpy
```
Comment (ebezzi) [inline backend/wmg/requirements.txt:82]:
```suggestion
```
Already defined at the end

**Review by ebezzi — APPROVED** [review]

File: backend/wmg/data/wmg_cube.py [hunk 2]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+def create_cubes(tdb_group):
+    create_expression_summary_cube(tdb_group)
+    create_cell_count_cube(tdb_group)
+
+
+def create_expression_summary_cube(tdb_group):
+    """
+    Create queryable cube and write to disk
+    """
+    uri = f"{tdb_group}/cube"
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:62]: not the name ultimately used by the API layer.  is it mapped later to correct name later on? if so, probably better to use the expected name throughout; we have a constant for it: backend/wmg/data/snapshot.py:17

File: backend/wmg/data/wmg_cube.py [hunk 3]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+        logger.debug("Cube consolidated, start vacuumming")
+        tiledb.vacuum(uri)
+
+    logger.debug("Cube creation complete")
+    create_cube_sec = time.time() - start_time
+    logger.info(f"Big cube: time to create {create_cube_sec}, uri={uri}")
+
+
+def create_empty_cube(uri: str):
+    """ "
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:88]:
```suggestion
    """
```

File: backend/wmg/data/wmg_cube.py [hunk 4]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+    Create an empty cube with expected schema (dimensions and attributes) at given uri
+    """
+    if tiledb.array_exists(uri):
+        tiledb.remove(uri)
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:92]: an overwrite flag is support by tiledb

File: backend/wmg/data/wmg_cube.py [hunk 5]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+    # Reduce X
+    ##
+    big_cube_atts = cube_indexed_dims_no_gene_ontology + cube_non_indexed_dims
+    cell_labels, cube_index = make_cube_index(tdb_group, big_cube_atts)
+    n_groups = len(cube_index)
+
+    cube_sum = np.zeros((n_groups, n_genes), dtype=np.float32)
+    cube_nnz = np.zeros((n_groups, n_genes), dtype=np.uint64)
+    cube_min = np.zeros((n_groups, n_genes), dtype=np.float32)
+    cube_max = np.zeros((n_groups, n_genes), dtype=np.float32)
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:122]: don't need min/min fwiw

File: backend/wmg/data/wmg_cube.py [hunk 6]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+
+    # pass 1 - sum, nnz, min, max
+    reduce_X(tdb_group, start_time, cell_labels.cube_idx.values, cube_sum, cube_nnz, cube_min, cube_max)
+
+    return build_in_mem_cube(gene_ontology_term_ids, cube_index, cube_non_indexed_dims, cube_sum, cube_nnz)
+
+
+def build_in_mem_cube(gene_ids, cube_index, other_attrs, cube_sum, cube_nnz):
+    """
+    Build the cube in memory, calculating the
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:132]: missing end of comment

File: backend/wmg/data/wmg_cube.py [hunk 7]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+        (
+            tissue_ontology_term_id,
+            organism_ontology_term_id,
+            *attr_values,
+            n,
+            cube_idx,
+        ) = grp.tolist()
+        mask = cube_nnz[cube_idx] != 0
+        n_vals = np.count_nonzero(mask)
+        if n_vals == 0:  # Todo what is the purpose of this?
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:168]: re:TODO, to maintain sparsity, presumably, if we want to update comment to that effect

File: backend/wmg/data/wmg_cube.py [hunk 8]
```diff
@@ -36,3 +48,216 @@ def create_cell_count_cube(tdb_group: str):
...
+    return dims, vals
+
+
+def reduce_X(tdb_group, start_time, cube_indices, *accum):
+    """
+    # TODO
+    """
+    with concurrent.futures.ThreadPoolExecutor() as tp:
+        cfg = {
+            "py.init_buffer_bytes": 512 * 1024**2,
```
Comment (atolopko-czi) [inline backend/wmg/data/wmg_cube.py:195]: we have a `MB` constant (1024**2)

**Review by atolopko-czi — APPROVED** [review]
Nuthin' but nits from me. 😄  Have a look at them, but otherwise LGTM, from an admittedly  quick review.

File: backend/wmg/data/wmg_cube.py (hunk 7 above)
Comment (MDunitz) [reply backend/wmg/data/wmg_cube.py:168]: thanks

File: backend/wmg/data/wmg_cube.py (hunk 2 above)
Comment (MDunitz) [reply backend/wmg/data/wmg_cube.py]: updated with the correct constant

File: backend/wmg/data/wmg_cube.py (hunk 5 above)
Comment (MDunitz) [reply backend/wmg/data/wmg_cube.py]: these are actually used by coo_cube_pass1_into so I need to keep them

**Pull request merged**